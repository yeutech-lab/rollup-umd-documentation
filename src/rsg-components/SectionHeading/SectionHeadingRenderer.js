import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import A from 'bootstrap-styled/lib/A';
import cn from 'classnames';
import omit from 'lodash.omit';
import mapToCssModules from 'map-to-css-modules/lib';
import Heading from '../Heading';

export const defaultProps = {
  theme: {
    styleguide: {
      '$rsg-section-heading-display': 'flex',
      '$rsg-section-heading-flex-direction': 'row',
      '$rsg-section-heading-align-items': 'center',
      '$rsg-section-heading-margin-bottom': '8px',
      '$rsg-section-heading-section-name-isolation': 'false',
      '$rsg-section-heading-section-name-text-decoration': 'underline',
      '$rsg-section-heading-section-name-cursor': 'pointer',
      '$rsg-section-heading-section-name-color': '#B31255',
      '$rsg-section-heading-deprecated-text-decoration': 'line-through',
      '$rsg-section-heading-deprecated-cursor': '#767676',
      '$rsg-section-heading-toolbar-margin-left': 'auto',
      '$rsg-section-heading-1-link-color': '#292b2c',
      '$rsg-section-heading-2-link-color': '#767676',
    },
  },
};

export const propTypes = {
  /**
   * @ignore
   */
  className: PropTypes.string, // eslint-disable-line react/require-default-props
  /** Specified node element will be passed as children of `<SectionHeadingRenderer />` component. */
  children: PropTypes.node, // eslint-disable-line react/require-default-props
  /** Toolbar element to be rendered. */
  toolbar: PropTypes.node, // eslint-disable-line react/require-default-props
  /** Set id used by `<Heading />` component. */
  id: PropTypes.string.isRequired,
  /** Set anchor link href used by `<A />` component. */
  href: PropTypes.string.isRequired,
  /** Set level used in `<Heading />` component. */
  depth: PropTypes.number.isRequired,
  /** Toggle deprecated style. */
  deprecated: PropTypes.bool, // eslint-disable-line react/require-default-props
  /** Theme variables. Can be: */
  theme: PropTypes.shape({
    styleguide: PropTypes.shape({
      '$rsg-section-heading-display': PropTypes.string,
      '$rsg-section-heading-flex-direction': PropTypes.string,
      '$rsg-section-heading-align-items': PropTypes.string,
      '$rsg-section-heading-margin-bottom': PropTypes.string,
      '$rsg-section-heading-section-name-isolation': PropTypes.string,
      '$rsg-section-heading-section-name-text-decoration': PropTypes.string,
      '$rsg-section-heading-section-name-cursor': PropTypes.string,
      '$rsg-section-heading-section-name-color': PropTypes.string,
      '$rsg-section-heading-deprecated-text-decoration': PropTypes.string,
      '$rsg-section-heading-deprecated-cursor': PropTypes.string,
      '$rsg-section-heading-toolbar-margin-left': PropTypes.string,
      '$rsg-section-heading-1-link-color': PropTypes.string,
      '$rsg-section-heading-2-link-color': PropTypes.string,
    }),
  }),
  /**
   * Replace or remove a className from the component.
   * See example <a href="https://www.npmjs.com/package/map-to-css-modules" target="_blank">here</a>.
   */
  cssModule: PropTypes.object, // eslint-disable-line react/require-default-props
};

const SectionHeadingRendererUnstyled = (props) => {
  const {
    className,
    cssModule,
    children,
    toolbar,
    id,
    href,
    depth,
    deprecated,
    ...attributes
  } = omit(props, ['theme']);

  const headingLevel = Math.min(6, depth);
  return (
    <div
      className={mapToCssModules(cn(className, 'rsg-section-heading'), cssModule)}
      {...attributes}
    >
      <Heading level={headingLevel} id={id}>
        <A className={`section-name level-${headingLevel} ${deprecated ? 'deprecated' : ''}`} href={href}>
          {children}
        </A>
      </Heading>
      <div className="toolbar">{toolbar}</div>
    </div>
  );
};

SectionHeadingRendererUnstyled.defaultProps = defaultProps;
SectionHeadingRendererUnstyled.propTypes = propTypes;

const SectionHeadingRenderer = styled(SectionHeadingRendererUnstyled)` 
  ${(props) => `
    &.rsg-section-heading {
      display: ${props.theme.styleguide['$rsg-section-heading-display']};
      flex-direction: ${props.theme.styleguide['$rsg-section-heading-flex-direction']};
      align-items: ${props.theme.styleguide['$rsg-section-heading-align-items']};
      margin-bottom: ${props.theme.styleguide['$rsg-section-heading-margin-bottom']};
      & .section-name {
        &:hover, &:active {
          isolation: ${props.theme.styleguide['$rsg-section-heading-section-name-isolation']};
          text-decoration: ${props.theme.styleguide['$rsg-section-heading-section-name-text-decoration']};
          cursor: ${props.theme.styleguide['$rsg-section-heading-section-name-cursor']};
          color: ${props.theme.styleguide['$rsg-section-heading-section-name-color']};
        }        
      }
      & .deprecated {
        text-decoration: ${props.theme.styleguide['$rsg-section-heading-deprecated-text-decoration']};
        cursor: ${props.theme.styleguide['$rsg-section-heading-deprecated-cursor']};
      }
      & .toolbar {
        margin-left: ${props.theme.styleguide['$rsg-section-heading-toolbar-margin-left']};
      }
      & .level-1 {
        color: ${props.theme.styleguide['$rsg-section-heading-1-color']};
      }
      & .level-2 {
        color: ${props.theme.styleguide['$rsg-section-heading-2-color']};
      }
    }
 `}
`;

SectionHeadingRenderer.defaultProps = defaultProps;
SectionHeadingRenderer.propTypes = propTypes;

export default SectionHeadingRenderer;
