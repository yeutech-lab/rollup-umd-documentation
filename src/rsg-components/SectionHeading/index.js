export { default } from '../SectionHeading/SectionHeading';
