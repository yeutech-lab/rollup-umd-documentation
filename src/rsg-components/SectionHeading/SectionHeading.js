import React from 'react';
import PropTypes from 'prop-types';
import Slot from 'rsg-components/Slot';
import SectionHeadingRenderer from 'rsg-components/SectionHeading/SectionHeadingRenderer';
import getUrl from '../../utils/getUrl';

export default function SectionHeading({ slotName, slotProps, children, id, ...rest }) {
	const href = getUrl({ slug: id, anchor: true });
	return (
		<SectionHeadingRenderer
			toolbar={<Slot name={slotName} props={slotProps} />}
			id={id}
			href={href}
			{...rest}
		>
			{children}
		</SectionHeadingRenderer>
	);
}

SectionHeading.propTypes = {
	children: PropTypes.node,
	id: PropTypes.string.isRequired,
	slotName: PropTypes.string.isRequired,
	slotProps: PropTypes.object.isRequired,
	depth: PropTypes.number.isRequired,
	deprecated: PropTypes.bool,
};
