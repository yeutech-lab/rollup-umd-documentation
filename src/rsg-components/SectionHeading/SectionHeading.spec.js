import React from 'react';
import SectionHeading from './index';
import SectionHeadingRenderer from './SectionHeadingRenderer';

describe('SectionHeading', () => {
	const FakeToolbar = () => <div>Fake toolbar</div>;

	it('should forward slot properties to the toolbar', () => {
		const actual = shallow(
			<SectionHeading id="section" slotName="slot" slotProps={{ foo: 1, bar: 'baz' }} depth={2}>
				A Section
			</SectionHeading>
		);

		expect(actual).toMatchSnapshot();
	});

	it('should render a section heading', () => {
		const actual = shallow(
			<SectionHeadingRenderer id="section" href="/section" depth={2} toolbar={<FakeToolbar />}>
				A Section
			</SectionHeadingRenderer>
		);

		expect(actual.dive()).toMatchSnapshot();
	});

	it('should render a deprecated section heading', () => {
		const actual = shallow(
			<SectionHeadingRenderer
				id="section"
				href="/section"
				depth={2}
				toolbar={<FakeToolbar />}
				deprecated
			>
				A Section
			</SectionHeadingRenderer>
		);

		expect(actual.dive()).toMatchSnapshot();
	});

	it('should prevent the heading level from exceeding the maximum allowed by the Heading component', () => {
		const actual = shallow(
			<SectionHeadingRenderer id="section" href="/section" depth={7} toolbar={<FakeToolbar />}>
				A Section
			</SectionHeadingRenderer>
		);

		expect(
			actual
				.dive()
				.find('Styled(Heading)')
				.prop('level')
		).toEqual(6);
	});
});
