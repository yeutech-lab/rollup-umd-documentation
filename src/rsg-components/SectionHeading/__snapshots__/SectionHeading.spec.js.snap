// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionHeading should forward slot properties to the toolbar 1`] = `
<Styled(SectionHeading)
  depth={2}
  href="blank#section"
  id="section"
  toolbar={
    <Slot
      name="slot"
      props={
        Object {
          "bar": "baz",
          "foo": 1,
        }
      }
    />
  }
>
  A Section
</Styled(SectionHeading)>
`;

exports[`SectionHeading should render a deprecated section heading 1`] = `
<div
  className="rsg--wrapper-0"
>
  <Styled(Heading)
    id="section"
    level={2}
  >
    <a
      className="rsg--sectionName-2 rsg--isDeprecated-3"
      href="/section"
    >
      A Section
    </a>
  </Styled(Heading)>
  <div
    className="rsg--toolbar-1"
  >
    <FakeToolbar />
  </div>
</div>
`;

exports[`SectionHeading should render a section heading 1`] = `
<div
  className="rsg--wrapper-0"
>
  <Styled(Heading)
    id="section"
    level={2}
  >
    <a
      className="rsg--sectionName-2"
      href="/section"
    >
      A Section
    </a>
  </Styled(Heading)>
  <div
    className="rsg--toolbar-1"
  >
    <FakeToolbar />
  </div>
</div>
`;
